### DISCUSSION TITLE
Minimum binary sizes on microcontrollers

### DISCUSSION THREAD

**Discussion by kbotteon** [nasa/fprime#1675] [Q&A]
For those of you who use FPrime on microcontrollers, what approximate binary sizes have you seen for minimal deployments, without the overhead of your application specific code?

I've heard of using FPrime on platforms with as little as 256 kiB of flash, but more commonly I see it used on platforms with 4 MiB or more available to store the program itself.

I'm trying to estimate the approximate minimum binary size for just the FPrime plumbing: PassiveComponentBase, Ports, Rate Groups, etc. to understand what is the floor on microcontroller non-volatile memory specs.

**Comment by SterlingPeet** [comment]
I cram F Prime onto a 128k program memory device with 64 kb of SRAM.  I’d say we get about 30 kb of application code on a 8 bit system if you want the nice services like commands, telemetry, and events.  We also have to use a number of code-reduction techniques, and top out around 10-12 components, total.

I would argue this is about as low as you can go, if you are asking for a self-contained system.  I am looking into smaller options if your deployment is essentially hung on the side of a bigger processor that supplies the nice services like command dispatching and telemetry reporting.

**Reply by kbotteon** [reply]
Nice! I'm planning to bump my system memory to 256 kiB to alleviate some risk of not fitting in a smaller memory.

>30 kb of application code on a 8 bit system if you want the nice services like commands, telemetry, and events

I didn't expect that. I recently tried building a minimal FPrime distribution with the platform vendor compiler and see a ~150 kiB binary when including those services. The symbol table is filled with every variety of printing and string formatting routines. Do you have text logging and asserts turned off in your FpConfig?

Thanks for sharing your experience!

**Reply by Joshua-Anderson** [reply]
@kbotteon Disabling [`FW_OBJECT_NAMES` in your config](https://github.com/nasa/fprime/blob/master/config/FpConfig.hpp#L91) should significantly reduce the number of strings in the binary

**Comment by SterlingPeet** [comment]
I turn these off as mentioned here.

Sterling

Sent from my mobile.

On Oct 4, 2022, at 2:11 PM, Joshua Anderson ***@***.***> wrote:

﻿

@kbotteon<https://github.com/kbotteon> Disabling FW_OBJECT_NAMES in your config<https://github.com/nasa/fprime/blob/master/config/FpConfig.hpp#L91> should significantly reduce the number of strings in the binary

**Comment by kbotteon** [comment]
I'll give that a try. Thank you both!

**Comment by ThibFrgsGmz** [comment]
@SterlingPeet I tried to see the minimum footprint of F' by playing around with disabling symbols but I admit to being surprised at what I find.
I defined a topology of 2 components which are empty and I can't get below 200 Kb.

By doing a lot of tests I may have lost my way, here is the repository where I did my tests: https://github.com/ThibFrgsGmz/fprime-mem-footprint-benchmark

Can you tell me what you think?

**Reply by SterlingPeet** [reply]
That’s probably about right for the configurations you specified.  We get smaller deployments because the metal is 8-bit, not 64-bit so we don’t have any wasted space between smaller words in our binaries.  We also disable 64bit types in F Prime (which is not as well tested as we would like, but when we run into bugs, we generally report them 😄).  We are also using the bare metal scheduler, so there might be more memory used for threading support.

**Reply by ThibFrgsGmz** [reply]
The amazing thing is that I only have two passive components in the binary: there is no syscall at all. Basically, it's just two instances of two different classes, one passing a float to the other.

I feel like I made a big mistake that I can't even see. I'll leave it at that for today on this benchmark, maybe I'll have better ideas later when I've rested.

**Reply by LeStarch** [reply]
Did you supply your own `new` and `delete` implementations?  We found that using the C++ built-in `new` and `delete` includes the exception library (large) which itself pulls in `printf` (larger).  These are included even when using new without exceptions as we are doing.

Define in a file linked to main a few functions that map `new` to `malloc` and we saved ~70KB of space.

**Reply by kbotteon** [reply]
One thing to check: Some embedded toolchains link libstdc++ **with** exception handling code regardless of compiler and linker settings because a static library compiled with exception handling is the only thing packaged with the toolchain. Since FPrime uses new, delete, and pure virtual functions, this can lead to the inclusion of bulky exception handling code in the final binary. A workaround is to redefine `new`, `delete` and `__cxa_pure_virtual` with implementations that cannot throw.

I've also seen a really good histogram showing code sizes for FPrime built-in Components floating around. I'll ask if that can be released for community reference. For instance, on my platform, `Fw::Types` required about ~50 KiB of text and data.

Edit: You beat me to it by a few seconds!

**Reply by LeStarch** [reply]
Another thing that can be done is to look through the output of `objdump` and see what the large masses of code are.   This is how we found the above issue: dumped the object, noticed `printf` despite cutting printf usage, traced the dependency, and finally removed the dependency. 

It is a time consuming process, but optimizing anything at the level may turn out time consuming.

**Reply by ThibFrgsGmz** [reply]
OK, thanks for the advice guys. I only built a `Ref` deployment stripped of components on my laptop/WSL-2. So the implementation of new/delete etc. comes from my GNU 10.x toolchain.

I'm really interested in your histograms @kbotteon !

Internally, I am often asked about the quality of the built-in components (coding rules compliance like AUTOSAR, 100% MC/DC coverage, size, speed, ...).

**Reply by astroesteban** [reply]
Side note: I came across [this post](https://interrupt.memfault.com/blog/code-size-deltas) from MemFault on tracking firmware size. You might get some good ideas from this @ThibFrgsGmz.

**Reply by ThibFrgsGmz** [reply]
OK I understood my mistake thanks to Google's [bloaty](https://github.com/google/bloaty) - thanks @astroesteban for the link!.

It gave me a size of 42.3Ki (almost 44 Kb) and that's when I understood the beginner's mistake I made...

Unlike `readelf` where the sizes are given in **hexa**, the `ll` command gives the sizes in **decimal**. So I get a minimal **binary** image of F´ with Linux/GNU 10.4 of about **32 KB**.

It's more reasonable, though maybe still pretty big for what it does. I'll look at it later to see what it really is.
